File: 프로그래머스/3/72414. 광고 삽입/광고 삽입.py
from collections import deque

def str_to_int(str_time):
    tmp=0
    for j,v in enumerate(list(map(int,str_time.split(":")))):
        tmp+=v*(60**(2-j))
    return tmp
def int_to_str(int_time):
    h = int_time//3600
    int_time%=3600
    m = int_time//60
    int_time%=60
    s = int_time
    return ':'.join([str(i).zfill(2) for i in [h,m,s]])

def solution(play_time, adv_time, logs):
    play_time = str_to_int(play_time)
    imos = [0]*(play_time+2)
    for i in logs:
        s,e=i.split('-')
        imos[str_to_int(s)]+=1
        imos[str_to_int(e)]-=1
    now = 0
    for i in range(len(imos)):
        now+=imos[i]
        imos[i]=now
    adv_time = str_to_int(adv_time)
    
    # print(imos[str_to_int("01:31:00")-4 : str_to_int("01:31:00")+3])
    # print(imos[str_to_int("01:31:00")+adv_time-4 : str_to_int("01:31:00")+adv_time+3])
    # print(str_to_int("01:31:00")-4, str_to_int("01:31:00")+3)
    ns = 0
    for i in range(adv_time):
        ns+=imos[i]
    s,e = 0, adv_time-1
    answer = [ns,s,e]
    cnt=0
    while e<len(imos)-1:
        ns-=imos[s]
        s+=1
        e+=1
        ns+=imos[e]
        
        if ns>answer[0]:
            cnt+=1
            answer = max(answer,[ns,s,e])
        
    
    
    return int_to_str(answer[1])

File: 프로그래머스/3/72414. 광고 삽입/광고 삽입.java
import java.util.*;

class Solution {
    public String solution(String play_time, String adv_time, String[] logs) {
        int playSeconds = strToInt(play_time);
        int advSeconds = strToInt(adv_time);
        long[] imos = new long[playSeconds + 2]; // 최대 360,000 초 이상

        for (String log : logs) {
            String[] times = log.split("-");
            int start = strToInt(times[0]);
            int end = strToInt(times[1]);
            imos[start] += 1;
            imos[end] -= 1;
        }

        // 누적 시청자 수 계산
        for (int i = 1; i < imos.length; i++) {
            imos[i] += imos[i - 1];
        }

        // 누적 재생시간 계산
        for (int i = 1; i < imos.length; i++) {
            imos[i] += imos[i - 1];
        }

        long maxTime = imos[advSeconds - 1];
        int startTime = 0;

        for (int i = advSeconds; i < playSeconds; i++) {
            long current = imos[i] - imos[i - advSeconds];
            if (current > maxTime) {
                maxTime = current;
                startTime = i - advSeconds + 1;
            }
        }

        return intToStr(startTime);
    }

    // HH:MM:SS -> 초(int) 변환
    private int strToInt(String time) {
        String[] parts = time.split(":");
        int h = Integer.parseInt(parts[0]);
        int m = Integer.parseInt(parts[1]);
        int s = Integer.parseInt(parts[2]);
        return h * 3600 + m * 60 + s;
    }

    // 초(int) -> HH:MM:SS 변환
    private String intToStr(int time) {
        int h = time / 3600;
        time %= 3600;
        int m = time / 60;
        int s = time % 60;
        return String.format("%02d:%02d:%02d", h, m, s);
    }
}
